ckanext/security/authenticator.py

```diff
@@ -132,6 +146,45 @@ def authenticate_totp(auth_user):
         return auth_user
 
 
+def login() -> Union[Response, str]:
+    """Override the CKAN default login functionality to provide
```

Comment:
Wouldn't it be simpler to just register the 2FA code as an authenticator that returns None on success and an error in any other case?

Reply:
Interesting idea... however, what I'm trying to achieve here is to make the login failure behaviour look the same regardless of the failure mode, so that malicious actors get no useful information about what the failure was. The default login() and ckan_authenticate() methods in CKAN core don't really seem to handle errors propagating from plugins that implement authenticate() via IAuthenticator, so returning an error doesn't appear like it would work anyway.

Please feel free to raise a PR if you have ideas about how this could be achieved without exposing information to attackers about what caused the auth to fail, i.e. login attempts after throttle lock OR with MFA failure should still display the standard login failure message.